src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs

```diff
@@ -305,38 +306,93 @@ protected override NamedTypeSymbol ContainingType
 
-                BoundNode scope;
-                if (!_analysis.VariableScope.TryGetValue(captured, out scope))
+                if (closure.MethodKind == MethodKind.LocalFunction &&
+                    OnlyCapturesThis(capturedVars))
                 {
                     continue;
                 }
 
-                LambdaFrame frame = GetFrameForScope(scope, closureDebugInfo);
-
-                if (captured.Kind != SymbolKind.Method)
+                foreach (var captured in capturedVars)
                 {
-                    var hoistedField = LambdaCapturedVariable.Create(frame, captured, ref _synthesizedFieldNameIdDispenser);
-                    proxies.Add(captured, new CapturedToFrameSymbolReplacement(hoistedField, isReusable: false));
-                    CompilationState.ModuleBuilderOpt.AddSynthesizedDefinition(frame, hoistedField);
+                    BoundNode scope;
+                    if (!_analysis.VariableScope.TryGetValue(captured, out scope))
+                    {
+                        continue;
+                    }
+
+                    LambdaFrame frame = GetFrameForScope(scope, closureDebugInfo);
 
-                    if (hoistedField.Type.IsRestrictedType())
+                    if (captured.Kind != SymbolKind.Method && !
+                        proxies.ContainsKey(captured))
                     {
-                        foreach (CSharpSyntaxNode syntax in kvp.Value)
+                        var hoistedField = LambdaCapturedVariable.Create(frame, captured, ref _synthesizedFieldNameIdDispenser);
+                        proxies.Add(captured, new CapturedToFrameSymbolReplacement(hoistedField, isReusable: false));
+                        CompilationState.ModuleBuilderOpt.AddSynthesizedDefinition(frame, hoistedField);
+
+                        if (hoistedField.Type.IsRestrictedType())
                         {
-                            // CS4013: Instance of type '{0}' cannot be used inside an anonymous function, query expression, iterator block or async method
-                            this.Diagnostics.Add(ErrorCode.ERR_SpecialByRefInLambda, syntax.Location, hoistedField.Type);
+                            foreach (CSharpSyntaxNode syntax in _analysis.CapturedVariables[captured])
+                            {
+                                // CS4013: Instance of type '{0}' cannot be used inside an anonymous function, query expression, iterator block or async method
+                                this.Diagnostics.Add(ErrorCode.ERR_SpecialByRefInLambda, syntax.Location, hoistedField.Type);
+                            }
                         }
                     }
                 }
             }
         }
 
+        /// <summary>
+        /// Helper for determining whether a local function transitively
+        /// only captures this (only captures this or other local functions
+        /// which only capture this).
+        /// </summary>
+        private bool OnlyCapturesThis<T>(
```

Comment:
It feels like this should take a method/closure symbol  as the input and have additional table on a side to keep previously computed resolutions, so that we would not compute this more than once per method/closure. 
Otherwise in a situation where N local functions call each other, the complexity would be exponential and that may become a problem even for a relatively small N.